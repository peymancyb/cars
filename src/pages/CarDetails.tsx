import React, {useCallback, useState, useEffect} from 'react';
import {useParams} from 'react-router-dom';
import Layout from '../components/Layout';
import CarApi from '../api';
import Button from '../components/Button';

interface ICar {
  stockNumber: number;
  manufacturerName: string;
  modelName: string;
  color: string;
  mileage: {
    number: number;
    unit: string;
  };
  fuelType: string;
  pictureUrl: string;
}

function CardDetails() {
  const [carDetails, setCarDetails] = useState<ICar | null>(null);
  const {stockNumber} = useParams();

  const getCar = useCallback(async () => {
    try {
      const {car} = await CarApi.getCarByStockNumber(stockNumber!);
      setCarDetails(car);
    } catch (error) {
      console.log('error -> ', error);
    }
  }, [stockNumber]);

  useEffect(() => {
    getCar();
  }, [getCar]);

  const getCarName = () => {
    const {manufacturerName, modelName} = carDetails!;
    return `${manufacturerName} ${modelName}`;
  };

  const getCarDetails = () => {
    const {mileage, fuelType, color} = carDetails!;
    return `Stock #${carDetails!.stockNumber} - ${
      mileage.number
    } ${mileage.unit.toUpperCase()} - ${fuelType} - ${color}`;
  };

  if (!carDetails) {
    return null;
  }

  return (
    <Layout>
      <div className="car-details-container">
        <div className="car-image-container">
          <img
            style={{marginLeft: 'auto', marginRight: 'auto', display: 'block'}}
            src={carDetails.pictureUrl}
            alt={`car-${carDetails.stockNumber}`}
          />
        </div>
        <div className="car-details-view">
          <div style={{width: 450}}>
            <p className="head-text">{getCarName()}</p>
            <p>{getCarDetails()}</p>
            <p>
              This car is currently available and can be delivered as soon as
              tomorrow morning. Please be aware that delivery times shown in
              this page are not definitive and may change due to bad weather
              conditions.
            </p>
          </div>
          <div className="save-box">
            <p>
              If you like this car, click the button and save it in your
              collection of favourite items.
            </p>
            <Button className="align-right" onPress={() => {}} text="Save" />
          </div>
        </div>
      </div>
    </Layout>
  );
}

export default CardDetails;
