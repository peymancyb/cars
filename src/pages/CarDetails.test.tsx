import React from 'react';
import {
  render,
  cleanup,
  waitForElement,
  waitForElementToBeRemoved,
} from '@testing-library/react';
import {Router} from 'react-router-dom';
import {createBrowserHistory} from 'history';
import CarDetails from './CarDetails';

afterEach(cleanup);

test('(CarDetails) component should match snapshot', () => {
  const history = createBrowserHistory();
  const {asFragment} = render(
    <Router history={history}>
      <CarDetails />
    </Router>,
  );
  expect(asFragment).toMatchSnapshot();
});

test('using "stockNumber" provided in url should get the car information if car is available', () => {
  const history = createBrowserHistory();
  history.push('/car/10000');
  const {getByTestId, getByText} = render(
    <Router history={history}>
      <CarDetails />
    </Router>,
  );

  return waitForElementToBeRemoved(() => getByText('Loading ...'))
    .then(() => {
      return waitForElement(() => getByText('Car not found!'))
        .then(() => {
          expect(getByText('Car not found!')).toBeTruthy();
        })
        .catch(() => {
          expect(
            getByText(
              'If you like this car, click the button and save it in your collection of favourite items.',
            ),
          ).toBeTruthy();
        });
    })
    .catch(err => {
      expect(err).toThrowError();
    });
});
