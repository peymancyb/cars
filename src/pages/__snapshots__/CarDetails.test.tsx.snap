// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`(CarDetails) component should match snapshot 1`] = `[Function]`;
